src/helpers/buffers/ordering_sender.rs

```diff
@@ -40,6 +40,8 @@ struct State {
     write_ready: Option<Waker>,
     /// Another entity to wake when the buffer is read from.
     stream_ready: Option<Waker>,
+    /// record whether the sender is idle recently
+    idle: bool,
```

Comment:
it is possible to generalize this by leveraging Rust's trait system

```rust
trait Observable {
    fn is_changed(&self) -> bool;
}

impl <K: Hash + Eq, V> Observable for DashMap<K, Watcher<V>> {
    fn is_changed(&self) -> bool {
        self.iter().any(|v| v.is_touched())
    }
}

/// Tracks whenever someone accesses `T` by a shared reference. If it happens, sets `accessed` to
/// `true` until someone calls `is_touched`.
struct Watcher<T> {
    inner: T,
    touched: Arc<AtomicBool>,
}

impl <T> Watcher<T> {
    fn watch(inner: T) -> Self {
        Self {
            inner,
            touched: Arc::new(AtomicBool::new(true)),
        }
    }
    
    fn is_touched(&self) -> bool {
        self.touched.swap(false, Ordering::Relaxed)
    }
}

impl <T> Deref for Watcher<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.touched.fetch_or(true, Ordering::Relaxed);
        &self.inner
    }
}

impl <T: Default> Default for Watcher<T> {
    fn default() -> Self {
        Self::watch(T::default())
    }
}

impl <T: Clone> Clone for Watcher<T> {
    fn clone(&self) -> Self {
        Self {
            inner: T::clone(&self.inner),
            touched: self.touched.clone(),
        }
    }
}
```

then you just implement it for senders and receivers

```rust
impl Observable for Watcher<GatewaySenders> {
    fn is_changed(&self) -> bool {
        self.is_touched() || self.inner.inner.is_changed()
    }
}
```

Reply:
Note that `Deref` allows us to track when someone calls `get_channel` or `send/receive` on that channel, but not the future poll like you implemented. However I think that may be enough because send and receive operations are cheap and happen many times per report/step.